DBZ-3382 Kafka sink adapter for Debezium Server

Alfusainey opened the pull request:
A kafka sink adapter for the debezium server. The Kafka producer currently requires only a single `bootstrap.servers` configuration property to find the kafka cluster to forward change events to.

This PR is currently a draft so that we can discuss possible configuration properties for the producer.

Signed-off-by: Alfusainey Jallow <alf.jallow@gmail.com>

jpechane commented:
@Alfusainey Great feature, thanks! May I ask you to modify the source config in the test to use the Kafka as the offset storage? I suppose this will be the mostly used configuration so it'd be greate to test it right away.

Alfusainey commented:
@jpechane Thanks a lot for the quick feedback!

> @Alfusainey Great feature, thanks! May I ask you to modify the source config in the test to use the Kafka as the offset storage? I suppose this will be the mostly used configuration so it'd be greate to test it right away.

Ok, i will add this in a subsequent commit.

> WRT the confgi options, please take a look at Pulsar. It uses just passthrough config so nee need to enumare the config options. I believe it can be the same for Kafka.

I just added a new commit to support passing the kafka configs through to the producer. can you take another look?

jpechane commented on debezium-server/debezium-server-kafka/pom.xml:
```diff
@@ -0,0 +1,91 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0"
```
I think this file use different indentation in comparison to opter poms.

jpechane commented on debezium-server/debezium-server-kafka/src/main/java/io/debezium/server/kafka/KafkaChangeConsumer.java:
```diff
@@ -0,0 +1,108 @@
+        final List<Future<RecordMetadata>> futures = new ArrayList<>();
+        for (ChangeEvent<Object, Object> record : records) {
+            try {
+                // TODO: change log level to trace
```
Please resolve the TODO

jpechane commented on debezium-server/debezium-server-kafka/src/main/java/io/debezium/server/kafka/KafkaChangeConsumer.java:
```diff
@@ -0,0 +1,108 @@
+            }
+        }
+
+        final List<Long> offsets = new ArrayList<>();
```
Is this only for logging? If yes then please all the protion of the code shoud be graded with `LOGGER.isTraceEnabled()`

Alfusainey replied:
I also used it as a way to `mark that the batch of records have been processed`. so yes, its not only for logging purposes. Or should we mark the batch as processed even though the batch may not have been sent to kafka?

jpechane replied:
It seems to me in the code that the variable is really used only in the log message.

Alfusainey replied:
if you're referring only to the list of offsets variable, yes that's true. But since getting the offsets will wait for the message to be committed in kafka, the last line `committer.markBatchFinished()` will IMO be executed only if all records are committed in kafka.  

If i would surround this in a logging enabled block and trace logging is not enabled, then a call to markBatchFinished() will not wait for all records to be committed in kafka.

But i will  update the patch as suggested

jpechane replied:
@Alfusainey I see now. I am sorry for the confusion as I mixed Connect offsets and kafka offsets together. So yes, it really makes sense to wait till the write is completed!
Then I'd still have a different proposal. Maybe you could use the send with callback method and use lath or barrier and wail for all to be completed instead of iteration over the list?

Alfusainey replied:
> Then I'd still have a different proposal. Maybe you could use the send with callback method and use lath or barrier and wail for all to be completed instead of iteration over the list?

yes, that make sense! ok, I will get back to this PR this weekend. The deadline for submitting the GSoC proposal is tomorrow and i have another deadline this Friday 😄

jpechane commented on debezium-server/debezium-server-kafka/src/test/java/io/debezium/server/kafka/KafkaITs.java:
```diff
@@ -0,0 +1,94 @@
+@QuarkusTest
+@QuarkusTestResource(PostgresTestResourceLifecycleManager.class)
+@QuarkusTestResource(KafkaTestResourceLifecycleManager.class)
+public class KafkaITs {
```
This should be `KafkaIT` not `KafkaITs`. Otherwise the test will not be executed.

jpechane commented on debezium-server/debezium-server-kafka/pom.xml:
```diff
@@ -0,0 +1,91 @@
+	<packaging>jar</packaging>
+
+	<properties>
+		<skipITs>true</skipITs>
```
This should be `false`. The tests should be executed by default as Kafka test can be soelf-contained.

jpechane commented:
@Alfusainey I've left couple more comments. Nothing too difficult. The only issue is that the IT test is not executed as part of build. When this is done, docuemntation is added and other comments addressed then I belive we are good to go.

Alfusainey commented:
> @Alfusainey I've left couple more comments. Nothing too difficult. The only issue is that the IT test is not executed as part of build. When this is done, docuemntation is added and other comments addressed then I belive we are good to go.

@jpechane thanks! I've addressed all the comments (will resolve them soon on GH) but am currently faced with an annoying issue: configuring the `KafkaProducer`  in the kafka change consumer. I tried configuring the `bootstrap.servers` config property in the test  config source as follows:

```Java
...
kafkaConfig.put("debezium.sink.kafka.producer.bootstrap.servers", KafkaTestResourceLifecycleManager.getBootstrapServers());
...
```

However, the problem is that quarkus first attempts to instantiate the `org.eclipse.microprofile.config.Config` object before the kafka container is started. I cannot pass a literal string value (e.g PLAINTEXT://localhost:9092) because the kafka port is dynamically assigned (and is different for each test run). I also tried setting it as a system property in `KafkaTestResourceLifecycleManager#start()` (after the container is started) but this does not work as well

Any hint on how i can workaround this?

jpechane commented on debezium-server/debezium-server-kafka/src/main/resources/META-INF/beans.xml:
```diff
@@ -0,0 +1 @@
+
```
Please remove the empty line

gunnarmorling commented:
Hey, so what's missing here to move forward with this one and un-draft it?

Alfusainey commented:
> Hey, so what's missing here to move forward with this one and un-draft it?

sorry, i haven't had time yet to take a look... only some minor changes are left + running the IT tests.  i will work on this on friday but should finish by the beginning of next week

Alfusainey commented:
@jpechane @gunnarmorling This PR is now ready for another review. I've incorprated all comments, ran ITs and added documentation

jpechane commented:
@Alfusainey Closing in favour of https://github.com/debezium/debezium/pull/2340 where I rebased the code and reformatted commit messages and applied. Thanks a lot for the contribution!
